### PR TITLE
Batch asyncindicators with a fix

### PR DESCRIPTION
**Pull request by sravfeyn**
Reverting https://github.com/dimagi/commcare-hq/pull/20947 with a [fix](https://github.com/dimagi/commcare-hq/pull/20949) for the bug because of which it was reverted.

### CODE REVIEW

**Comment by sravfeyn** [comment]
@emord @nickpell

**Review by nickpell — APPROVED** [review]

**Review by emord — COMMENTED** [review]
I'm not sure I fully understand how this issue occurred on ICDS.

Where did the `[]` come from? Everytime I try to step through this code, I can't figure it out

File: corehq/apps/userreports/tests/test_async_indicators.py [hunk 1]
```diff
@@ -69,3 +75,175 @@ def test_async_configured(self):
...
+        AsyncIndicator.objects.all().delete()
+        for adapter in self.adapters:
+            adapter.clear_table()
+
+    def indicators(self):
+        return AsyncIndicator.objects.all()
+
+    def _assert_rows_in_ucr_table(self, config, rows):
+        results = list(load_data_from_db(get_table_name(self.domain.name, config.table_id)))
+        if not rows:
```
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/userreports/tests/test_async_indicators.py]: There would be a `KeyError` in following lines as it refers to `rows[0]`

**Comment by sravfeyn** [comment]
> I'm not sure I fully understand how this issue occurred on ICDS.
> 
> Where did the [] come from? Everytime I try to step through this code, I can't figure it out

If all the indicators are case-type and one of the config is on form-type, then all the to be saved rows returned by `get_all_values` will be an empty set for this particular config. The empty set `[]` comes from [here](https://github.com/dimagi/commcare-hq/pull/20964/files#diff-1f15a4ea1fec5c4ee5fa155d4d80c3d7R342) from the fact that that we don't check for `config.filter(doc)` before calling `get_all_values`.

**Comment by sravfeyn** [comment]
> Everytime I try to step through this code, I can't figure it out

This is pretty much exactly inline with the pseudo-code you had, but most of the complicatedness comes from the fact that almost every statement has an individual try...except clause.

**Comment by emord** [comment]
@sravfeyn  I think that there's something else going on here. that this won't solve. Looking at https://sentry.io/dimagi/commcarehq/issues/580399479/events/22764852212/?environment=icds-new

One of the cases in this bulk operation was [de839c51-d60a-4b06-8da2-581027c40eea](https://www.icds-cas.gov.in/hq/admin/raw_doc/?id=de839c51-d60a-4b06-8da2-581027c40eea) which is a ccs_record. I used https://www.icds-cas.gov.in/a/icds-cas/configurable_reports/data_source_debugger/ to verify that this record should have rows written to the data source, but in the sentry error `rows = []`. To me this means that it wasn't going to write anything to that table for those indicators and effectively not update with the latest information. Does that sound like what is happening?

**Comment by sravfeyn** [comment]
Yeah, I think there is more to it. I ran the code here that calculates data against the indicators from the sentry error, and it returns non-empty set of rows. So, basically the same code returned empty then, but not now.

The fix I made should still be here since it addresses the problem of sqlalchemy error when trying to save empty rows, but yeah, this other problem needs more digging.

**Comment by emord** [comment]
> Yeah, I think there is more to it. I ran this code that calculates data in icds django shell (for the indicators from the sentry error), and it returns non-empty set of rows. So, basically the same code returned empty set then, but not anymore.

How did you test it? Did you test the sql adapter's `get_all_values` or did you run it through the task `build_async_indicators` ? My thought was that there is a bug in the task that was causing anything calculated by the adapter to be lost.

**Comment by emord** [comment]
@sravfeyn I had to roll this back. It nearly doubles the memory requirements for these celery machines which resulted in a lot of OOM in the queues.

You can see the difference in memory in all hte celery servers similar to this one 
https://app.datadoghq.com/dash/host/417532301?live=false&from_ts=1530033905461&to_ts=1530039167000&page=0&is_auto=false&tile_size=m

**Comment by sravfeyn** [comment]
@emord Thanks for testing. 

Where did the OOMerrors come from, is it from the new chunked processing code or sqlagg saving large number of rows or the way celery works (perhaps it doesn't release memory after the task ends)?

Do you think it's the code that's causing OOM errors? I have tested the code with a memory profiler to see any memory leaks or high memory usages (I've used as high as 1000 a chunk) and haven't found any bottlenecks. I have scanned through the task looking for big variables, the only big variable is the list of rows to be saved. 

What do you think is a next step?

**Comment by sravfeyn** [comment]
Did the older bug reappeared when you tested (the latest even of that bug on sentry is from the first time)?

**Comment by emord** [comment]
The older bug did not appear and it was fixed. The memory issue was the only issue I saw.

Looking at celery6 at the moment, it looks like each celery queue is taking ~6-7% of memory, but when I deployed this each was taking 12-15% of memory. There were a few UP/Down alerts because either the ucr queue or the reminder case update queues were being killed for low memory

I'm also not sure where the extra memory is coming from. I expected this to take more memory but not this much. Looking through the code I also couldn't tell what it would be. Before this change we already pulled from the database in bulk so this should only add the rows to save in memory. Quite confusing.

I think the next step may be trying to reduce batch size to 10 or something smaller to see if that helps. Then I can test that on deploy Monday/Tuesday

**Comment by emord** [comment]
I tried testing this out on one server today, but ran into some issues. First we'll want something like https://github.com/dimagi/commcare-hq/commit/5a881ef76330eb7c03964f206320a1a454da3210 to make sure we don't lose everything that's already queued. and I think we'll want to drop the assert https://sentry.io/dimagi/commcarehq/issues/618015330/?referrer=slack

Also I think that I was confused about what this chunk size was. I was expecting that it would be possible to pass in 100 documents in the function, but every 10 documents it would save. It looks like this is actually making a celery task for every 10 documents. Is it possible to decouple these numbers? (not sure if there are any issues creating 10x the celery tasks)

**Comment by sravfeyn** [comment]
Thanks @emord. I have [updated](https://github.com/dimagi/commcare-hq/pull/20964/commits/12fd5606f30fb7ecbe88df53015c5272529b968a) the correct chunk. Hope this is what you meant by `make sure we don't lose everything that's already queued`.

**Comment by sravfeyn** [comment]
Why to remove the assert that makes sure the input doesn't exceed `ASYNC_INDICATOR_CHUNK_SIZE`? I have removed it in anyway for now to unblock testing.

**Comment by emord** [comment]
I tested this out on celery7 and it doesn't look like it has any memory issues anymore. Can you resolve the conflicts with master, and implement something like https://github.com/dimagi/commcare-hq/commit/5a881ef76330eb7c03964f206320a1a454da3210 (so that we don't have a lot of errors from the `save_document` tasks already in the queue)

**Comment by sravfeyn** [comment]
> and implement something like 5a881ef

Do you mean like [this](https://github.com/dimagi/commcare-hq/pull/20964/commits/327e7169ce8bb8c2f0a721cba6515fc5dd4a8abc)?

@emord Also fixed the merge conflicts. Thanks for the tests!

**Comment by sravfeyn** [comment]
Couldn't reproduce the build fail locally, kicked build again. @emord If it goes well and if no other feedback, can you merge it today please.

**Comment by emord** [comment]
I added https://github.com/dimagi/commcare-hq/pull/20964/commits/ceae2d6bd1b5f3c72817a9aceaf89e3d53b939f2 so that `save_document` tasks that are queued will not cause errors. should be good to go after tests pass

**Comment by emord** [comment]
```

======================================================================
FAIL: corehq.apps.userreports.tests.test_pillow:AsyncIndicatorTest.test_async_save_success
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tests/test_pillow.py", line 437, in test_async_save_success
    self.assertEqual(rows.count(), 1)
AssertionError: 0L != 1
```
Is this the failure you mentioned @sravfeyn

**Comment by sravfeyn** [comment]
I made a fix, it was caching issue during tests.

**Pull request merged**